src/ZODB/Connection.py

```diff
@@ -1270,3 +1285,52 @@ def __repr__(self):
 large-record-size option in a configuration file) to specify a larger
 size.
 """
+
+class TransactionMetaData(object):
+
+    def __init__(self, user=u'', description=u'', extension=b''):
+        self.user = user
+        self.description = description
+        if not isinstance(extension, dict):
+            extension = _compat.loads(extension) if extension else {}
+        self.extension = extension
+
+    @property
+    def user(self):
+        return self.__user
+
+    @user.setter
+    def user(self, user):
+        if not isinstance(user, bytes):
+            user = user.encode('utf-8')
+        self.__user = user
+
+    @property
+    def description(self):
+        return self.__description
+
+    @description.setter
+    def description(self, description):
+        if not isinstance(description, bytes):
+            description = description.encode('utf-8')
+        self.__description = description
+
+    def note(self, text): # for tests
+        text = text.strip()
+        if not isinstance(text, bytes):
+            text = text.encode('utf-8')
+        if self.description:
+            self.description = self.description.strip() + b' ' + text
+        else:
+            self.description = text
+
+
+    @property
+    def extension(self):
+        return self.__extension
+
+    @extension.setter
+    def extension(self, v):
+        self.__extension = v
+
+    _extension = extension
```

Comment:
I'm not following your comment on copyTransactionsFrom.  BaseStorage's implementation, which uses copy, passes the iterator's transaction record to tpc_begin, which copies extension data.